ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java

```diff
@@ -1308,6 +1337,7 @@ public boolean isForgotten() {
 	
 	private TransactionImple _theTransaction;
     private boolean _recovered = false;
+	private boolean _rolledBack = false;
```

Comment:
I am not sure if it's important or not but I do not think you are making this part of the persistent state of the object so after recovery I suppose an extra rollback could be called?

Reply:
Yep, I also think that a situation like:
```
Prepare -> rollback (from us or from the resource) -> crash
```
can be a problem. Basically, if the RM initiated/received a rollback  but `_rolledBack` was not set to true, we will request rollback again on the same RM when we recover. However, even if we put some code in `save_state`/`restore_state` to try to save `_rolledBack` in the object store, there is still a chance to crash before  saving that variable...and we will need to request rollback again on the same resource. Having said that, the scenario where we request rollback again is not a problem as we should receive an `XAER_NOTA`, which we already handle correctly. Nevertheless, I introduced some code to try to save `_rolledBack` in the object store to reduce the chances to request rollback again after crashes.

Reply:
Thank you (BTW I am going to respond with more discussion but inline there rather than here and will resolve this thread)